AET-23: Aether ERD, entity extraction

quesebifurcan opened the pull request:
- Change gather2 models to reflect the aether ERD
- Add entity extraction utilities

Commit:
Changed references

Commit:
implemented new ERD

Commit:
removed signals import

Commit:
Fix: ran migrations for ODK importer

Commit:
fix: removed reference to survey in sezializers.py'

Commit:
fix: reference to survey in XFormAdmin

Commit:
refactored test_utils code

Commit:
added name field to models"

Commit:
updated: reference change in ODK and commmons module

Commit:
changed id to uuid4 for entity and response

Commit:
fixed linting error

Commit:
updated views tests and fixed double entity uuid generation

Commit:
exposed id field in entity for override

Commit:
Let response field in posted Entity be null, as not all resolve to a linked response

Commit:
stubbed out new functions for action dispatch in extractor

Commit:
wired up new dispatch function with UUID working and entity-reference as a stub

Commit:
type

Commit:
changed the id field in extraction from payload['_id'] to payload['id'] to get away from couchb convention

Commit:
working as intended

Commit:
fixed 400 error on ODK form submission

Commit:
fix: getting the mapping id now

Commit:
entity extraction is working, but only Person entities are being saved. Looking into it

Commit:
fixed a bug in the extractor that would only allow generation of one entity type per response

Commit:
small clean up

quesebifurcan committed:
chore: linting

quesebifurcan committed:
fix(core): tests

All "extractor actions" (gather2.core.api.utils.extractor_action)
are prefixed by "#!". Change test data to reflect this.

Fix typo in serializer: "revision" -> "map_revision"

Fix test for utils.get_entity_definitions() -- this function
now expects multiple schemas as its second parameter.

quesebifurcan committed:
fix: update request to core

quesebifurcan committed:
decrease coverage requirement; skip couchdb- and ui-tests

quesebifurcan committed:
skip view submission tests

quesebifurcan committed:
odk: linting

quesebifurcan committed:
chore: update requirements

Commit:
added modified field to Entity table

admbtlr commented on gather2-common/gather2/common/core/utils.py:
```diff
@@ -71,32 +71,32 @@ def check_connection():
 
 
-def get_surveys_url(survey_id=''):
+def get_surveys_url(mapping_id=''):
```
We should probably change this function name too

realdouno replied:
Fixed

admbtlr commented on gather2-common/gather2/common/core/utils.py:
```diff
@@ -71,32 +71,32 @@ def check_connection():
 
 
-def get_survey_responses_url(survey_id, response_id=None):
+def get_survey_responses_url(mapping_id, response_id=None):
```
Also this function name (i.e. `get_mapping_responses_url`). In fact actually we might want to review the nomenclature here, since you don't really "respond" to a mapping, but you could "submit". So ideally this would be `get_mapping_submissions_url`, but I guess that would involved a whole load of other changes.

realdouno replied:
`get_survey_responses_url` has been changed to `get_mapping_responses_url`. We can discuss this change in nomenclature during the standup

admbtlr commented on gather2-common/gather2/common/core/utils.py:
```diff
@@ -119,11 +119,11 @@ def get_data(url):
     '''
-    if survey_id is None:
+    if mapping_id is None:
         raise Exception('Cannot submit response without survey!')
```
Change this error message

realdouno replied:
Fixed

admbtlr commented on gather2-core/gather2/core/api/models.py:
```diff
@@ -1,99 +1,98 @@
 
-    code = models.TextField()
-    created = models.DateTimeField(auto_now_add=True, db_index=True)
+class Response(models.Model):
```
`Submission`?

realdouno replied:
Just to clarify, are you wanting us to replace `Response` by `Submission`?

admbtlr commented on gather2-core/gather2/core/api/utils.py:
```diff
@@ -51,3 +56,345 @@ def json_printable(obj):
+
+def get_entity_definitions(mapping_definition, schemas):
+    required_entities = {}
+    mapping = mapping_definition
```
Is this necessary?

realdouno replied:
Not really! Removed it

admbtlr commented on gather2-core/gather2/core/api/utils.py:
```diff
@@ -51,3 +56,345 @@ def json_printable(obj):
+
+
+def action_none():
+    # Called via #!none acts as a dummy instruction when you need to the first
```
Can you make this comment make sense?

realdouno replied:
@shawnsarwar

admbtlr commented on gather2-core/gather2/core/api/utils.py:
```diff
@@ -51,3 +56,345 @@ def json_printable(obj):
+        finally:
+            return value
+    else:
+        # create as little as the heirarchy as required
```
"make the hIErarcy as simple as possible"?

realdouno replied:
@shawnsarwar can confirm

admbtlr commented on gather2-core/gather2/core/api/utils.py:
```diff
@@ -51,3 +56,345 @@ def json_printable(obj):
+        source_data,
+):
+    # Takes an extractor action instuction (#!action#args) and dispatches it to
+    # the proper funciton
```
funcTIon

admbtlr replied:
instRuction

realdouno replied:
Fixed

admbtlr commented on gather2-core/gather2/core/api/utils.py:
```diff
@@ -51,3 +56,345 @@ def json_printable(obj):
+    elif action == "constant":
+        return action_constant(args)
+    else:
+        raise ValueError("No action by name %s" % action)
```
No action WITH name

realdouno replied:
Fixed

admbtlr commented on gather2-core/gather2/core/api/utils.py:
```diff
@@ -51,3 +56,345 @@ def json_printable(obj):
+    data = None
+    if response_data:
+        data = response_data
+    data["aether_errors"] = []
```
This is going to throw a TypeError if there is no `response_data`

realdouno replied:
`response_data` is assigned  `response.payload` and the payload is required when posting a  response so it's unlikely that we have no `response_data`. But I have set `data` to an empty dictionary in case there is no `response_data`

admbtlr commented on gather2-core/gather2/core/api/utils.py:
```diff
@@ -51,3 +56,345 @@ def json_printable(obj):
+    # For output. Since we need to submit the extracted entities as different
+    # types, it's helpful to seperate them here
+    entities = {}
+    # entitie names that have requirements
```
entitY

realdouno replied:
Fixed

admbtlr commented on gather2-core/gather2/core/api/utils.py:
```diff
@@ -51,3 +56,345 @@ def json_printable(obj):
+        raise ValueError("No action by name %s" % action)
+
+
+def extract_entity(requirements, response_data, entity_stubs):
```
This is a huge function - can we break it up?

admbtlr commented on gather2-core/gather2/core/api/utils.py:
```diff
@@ -51,3 +56,345 @@ def json_printable(obj):
+    entities = {}
+    # entitie names that have requirements
+    required_entities = requirements.keys()
+    for entity_name in required_entities:
```
I know that this is not the first time this terminology has been used, but I think we should probably call this `entity_type`. We're in danger of confusing entities with entity definitions, and I think we should do everything we can to make the distinction clearer. 

(Years ago, I worked as a developer and then trainer at a CMS company; our CMS had "items" and "items types", and it caused endless confusion for all our users)

admbtlr commented on gather2-core/gather2/core/api/utils.py:
```diff
@@ -51,3 +56,345 @@ def json_printable(obj):
+            # Some paths will satisfy more than one entity, so we increment in
+            # different places
+            i = 0
+            while i <= count-1:
```
why not `while i < count`?

admbtlr commented on gather2-core/gather2/core/api/utils.py:
```diff
@@ -51,3 +56,345 @@ def json_printable(obj):
+                if "#!" not in path:
+                    # Find the matches and assign them
+                    matches = parse(path).find(data)
+                    if len(matches) < 2:
```
Why not `if len(matches) == 1`? If there are 0 matches, line 305 will blow up.

admbtlr commented on gather2-core/gather2/core/api/utils.py:
```diff
@@ -51,3 +56,345 @@ def json_printable(obj):
+    # schema = project_schema.schema.definition
+
+    # Get entity definitions
+    entities = get_entity_definitions(mapping_definition, schemas)
```
Again here, we should always try and use `entity_defs` or similar

admbtlr commented on gather2-core/gather2/core/api/utils.py:
```diff
@@ -51,3 +56,345 @@ def json_printable(obj):
+    data, entities = extract_entity(requirements, response_data, entities)
+
+    entity_list = []
+    for name, entity_instances in entities.items():
```
See? now you have to use `entity_instances` because you used `entities` earlier :)

admbtlr commented on gather2-core/gather2/core/apps.py:
```diff
@@ -3,8 +3,8 @@
     def ready(self):
         ''' activates signals!!! '''
-        from .api import signals  # flake8: noqa
+        # from .api import signals  # flake8: noqa
```
delete

realdouno replied:
Done

admbtlr commented on gather2-core/gather2/core/tests/test_apps.py:
```diff
@@ -6,4 +6,4 @@ class AppsTests(TestCase):
 
     def test_app_config(self):
         # this is only valid in tests, the correct name is `gather2.core`
```
delete

realdouno replied:
This test for the verbose_ name of the app. I am not sure why we need to delete it

admbtlr commented on gather2-odk-importer/gather2/odk/api/core_utils.py:
```diff
@@ -0,0 +1,28 @@
+import os
```
So this is the file you had to create because the `common` module changes weren't being picked up, right? We need to sort that out...

quesebifurcan replied:
The common module changes will be picked up if you run `./scripts/upgrade_all.sh` and rebuild all containers (that will take a while).

realdouno replied:
Fixed

admbtlr reviewed (changes requested):
We should figure out a strategy for refactoring this PR. I just discussed it with Fredrik, and we're going to try and merge his fixed tests into this branch, so that we can then refactor in confidence (TM)

quesebifurcan committed:
Mapping ids, test coverage, couchdb-sync integration (#5)

gather2-common: Update submit_to_core() to use payload instead of data.

gather2-core: Reset migrations.

gather2-core: Fix inconsistencies in routes.
Resolves https://jira.ehealthafrica.org/browse/AET-44

gather2-core: Change CharField to UUIDField for all models

gather2-odk-importer: Update all tests to use UUIDs

tests: Use fixtures to bootstrap example project setups. odk-importer uses an empty schema (does not validate responses); couchdb-sync validates responses.

tests: Bring coverage back up to 82, 98 and 100 percent for core, odk-importer and couchdb-sync, respectively.

Commit:
refactored aether-23 based on comments

admbtlr reviewed (approved):
Pending merge of Douno's branch

quesebifurcan committed:
handle case with no data matching path

quesebifurcan committed:
Skip randomly failing couchdb test

quesebifurcan committed:
Couchdb coverage 100%
